### ISSUE TITLE
Client: review work request issues following #2918 and #3001

### ISSUE THREAD

**Issue by RichardHaselgrove** [BOINC/boinc#3065]
#2918 (fix job scheduling bug) has been accepted and merged. But in the process, adjustments were made to work fetch which were less than ideal. Quoting from the PR,

https://github.com/BOINC/boinc/pull/2918#issuecomment-450718595: @davidpanderson wrote

> I'm going to stop here with this PR. Maintaining a full buffer in the presence of max_concurrent would require nontrivial changes to the RR simulation which I might do later.

https://github.com/BOINC/boinc/pull/2918#issuecomment-460804035: @JuhaSointusalo wrote

> Even though I merged this I don't consider the job done yet. As previously discussed in my opinion the work fetch code now needs improving.

- If work fetch is blocked as long as there are max_concurrent or more tasks in buffer then before work fetch resumes there has to be an idle device.
- Those who want to use max_concurrent will now have to choose between using max_concurrent and having work buffered. My gut feeling is that people will choose work buffer and that means they won't be able to upgrade BOINC beyond 7.14 as long as the work fetch works the way it does now.

My view remains that the work fetch ceiling should, as now, reflect the sum of the base and additional work buffers, expressed in wall-clock time but proportional to the number of devices in use (max_concurrent rather than nCPUs).

#3001 (fix possible overflow in peak FLOPS) has also been merged, and also fixed in server release 1.0.4. But in [Message Board conversation](https://boinc.berkeley.edu/forum_thread.php?id=12842&postid=90266#90266) I noticed an anomaly, not previously reported on GitHub.

Work fetch is currently determined by [Client Scheduling October 2010 - Estimated credit](https://boinc.berkeley.edu/trac/wiki/ClientSchedOctTen#Estimatedcredit). I compared two separate work fetch logs, the first with the Peak FLOPS overflow bug, and the second without.

```
				RAC		    REC
Einstein@Home		  18,906.01	142,297,886.589
MilkyWay		   4,143.10	    216,419.883
SETI@home		   6,462.70	605,565,081.292

OpenCL: AMD/ATI GPU 0: AMD Radeon(TM) Vega 8 Graphics (43,980,464 GFLOPS peak)
```
The SETI GPU REC has been blown out of the water by the GFlops error. I don't think we'd thought of that before. If that's the only GPU in the system, it shouldn't make much difference - but it would unbalance the work fetch calculations between CPU, other GPUs, and the Ryzen GPU component.
```
				RAC		    REC
Einstein@Home		   8,119.18	      8,322.576
GPUGRID			 391,764.22	    303,164.317
NumberFields@home	   3,222.80	      1,834.406
SETI@home		  10,627.91	    367,015.648

OpenCL: NVIDIA GPU 0: GeForce GTX 970 (4,087 GFLOPS peak)
OpenCL: NVIDIA GPU 1: GeForce GTX 750 Ti (1,639 GFLOPS peak)
OpenCL: Intel GPU 0: Intel(R) HD Graphics 4600 (192 GFLOPS peak)
```
This is one of my machines, and I run:

- Einstein on the Intel GPU
- GPUGrid on the GTX 970
- SETI on the GTX 750 Ti and (rarely) on the GTX 970
- NumberFields on the CPU

Given the relative speeds of the two NVidia cards, the REC for SETI should be not much over half the REC for GPUGrid: instead, it's higher. That suggests that REC is being calculated from 'best card in system' (as reported to the server), rather than 'card actually used' (which should be knowable by the client). Again, future work requests could be unbalanced.

**Comment by davidpanderson** [comment]
Is there a client simulator scenario that shows a/this problem?

**Comment by RichardHaselgrove** [comment]
I haven't uploaded one yet, but I can try to do so if it would help. Which aspects of the work would a simulation be most useful for?

If it's the #2918 work fetch problem, it would probably be best if @KeithMyers uploaded one - he experienced the problem most severely with his more powerful machines. I could do a more limited set in the morning - getting late here..

**Comment by KeithMyers** [comment]
So what commit should I be running for the simulator?

**Comment by davidpanderson** [comment]
Unless you're building the simulator yourself, there's no choice.
It's generally based on master, except during periods when I'm working on it.

**Comment by KeithMyers** [comment]
In my https://github.com/BOINC/boinc/pull/2918 request, Richard had me compile a specific commit in Linux.  I was asking which commit version you wanted me to compile for testing on the simulator.

**Comment by RichardHaselgrove** [comment]
No, David is right. You only supply the data files for the run. The simulator uses its own executable files (I'd guess Linux builds). The simulator will already have the #2918 patch, because those changes have already been merged into master (as @JuhaSointusalo's comment confirms), and it'll also have #3001.

What you are requested to include, please, are the full set of app_config.xml files which caused problems with the test builds.

**Comment by RichardHaselgrove** [comment]
I have uploaded scenario 171 to explore the REC calculation on devices of different speeds. GPUGrid runs on NVidia Device 0 (excluded from Device 1): SETI runs on NVidia Device 1. REC half life is 1 day. If individual device speeds are taken into account, GPUGrid REC should be more than double SETI (in this scenario, where no manual intervention is possible).

https://boinc.berkeley.edu/scenarios/171/simulations/0/rec.png

**Comment by KeithMyers** [comment]
OK, yes I understand now you have already merged the commits you wanted tested.  I can supply files for the new scenario.

**Comment by RichardHaselgrove** [comment]
Well, it was Juha that merged them - BOINC has adopted a 'four eyes' approach, with separate committers and mergers, to reduce the risk of typos slipping through. But the merge was done under caution (as quoted above), because of this additional work still being needed.

**Comment by KeithMyers** [comment]
Hmmm, I attempted to create a scenario.  Uploaded all the files but no scenario has appeared.  Didn't see any warning messages or errors.

**Comment by AenBleidd** [comment]
Fixed via #4592
